File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        Deque<Character> stack = new ArrayDeque<>();
            
        stack.push(number.charAt(0));
        
        for (int i = 1; i < number.length(); i++) {
            while (!stack.isEmpty() && k > 0 && stack.peek() < number.charAt(i)) {
                stack.pop();
                k--;
            }
            stack.push(number.charAt(i));
        }
        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        String answer = sb.reverse().toString();
        
        return k == 0 ? answer : answer.substring(0, answer.length() - k);
    }
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
from collections import deque

def solution(number, k):
    q = deque()
    
    i = 0
    while k > 0 and i < len(number):
        while q and number[i] > q[-1] and k > 0:
            q.pop()
            k -= 1
        q.append(number[i])
        i += 1
    
    if k == 0:
        return ''.join(q) + ''.join(number[i:])
    
    return ''.join(q)[:-k]